vs-1997

AndriiSobko opened the pull request:
Unittests will be added in separate ticket.

AndriiSobko committed:
VS-1997

added base functionality for xlsx reporting

Commit:
Half-adapted this to xlsxwriter

AndriiSobko committed:
VS-1997

added create dropdown feature

AndriiSobko committed:
VS-1997

added conditional format, dropdown, docstrings

AndriiSobko committed:
VS-1997

added conditional format, dropdown, docstrings

AndriiSobko committed:
VS-1997

added conditional format, dropdown, docstrings

AndriiSobko committed:
VS-1997

added conditional format, dropdown, docstrings

rklyne commented on tablib/formats/_xlsx.py:
```diff
@@ -5,22 +5,21 @@
 
 import sys
 
+import openpyxl
```
Is this being added back because the removal of this code is not complete yet?

AndriiSobko replied:
this import was moved from line 14

rklyne commented on tablib/formats/_xlsx.py:
```diff
@@ -104,46 +154,25 @@ def import_book(dbook, in_stream, headers=True):
 
 
-def dset_sheet(dataset, ws, freeze_panes=True):
+def dset_sheet(dataset, ws, cell_formats):
```
We would probably need to add `freeze_panes` back in before this would be accepted upstream

AndriiSobko replied:
I can restore it, but, currently, this parameter isn't used in the function

rklyne commented on tablib/core.py:
```diff
@@ -452,6 +482,164 @@ def load(self, in_stream, format=None, **kwargs):
+        :param column_name: column name, str
+        :return: column index, int
+        """
+        if column_name:
```
Why would column_name not be provided?

AndriiSobko replied:
I will remove this check

AndriiSobko replied:
done

rklyne commented on tablib/core.py:
```diff
@@ -452,6 +482,164 @@ def load(self, in_stream, format=None, **kwargs):
+        :param aligment: text position in the cell, str
+        :return: dict
+        """
+        return {
```
Future improvement - make this a `Format()` object to support future improvements without changing the API.

AndriiSobko replied:
Cool, let's discuss it internally

rklyne commented on tablib/core.py:
```diff
@@ -452,6 +482,164 @@ def load(self, in_stream, format=None, **kwargs):
+        return {
+            'first_row': first_row,
+            'first_col': first_col,
+            'last_row': last_row,
```
Do we really have to store last_row and last_col ?
My concern here is that those change as the sheet grows.

AndriiSobko replied:
yes, we need those values in data_validation and condition_format methods:
https://xlsxwriter.readthedocs.io/worksheet.html#data_validation
https://xlsxwriter.readthedocs.io/worksheet.html#conditional_format
because we are working with the range of the cells

rklyne commented on tablib/core.py:
```diff
@@ -452,6 +482,164 @@ def load(self, in_stream, format=None, **kwargs):
+        for condition in conditions:
+            cell_format = {'bg_color': condition.get('color', None)}
+            value = '"{}"'.format(condition['value']) if isinstance(condition['value'], str) else condition['value']
+            self._set_conditional_format(1, column_position, len(self._package(dicts=False)) - 1,
```
We should avoid generating `_package` here - it's an expensive operation.
`self.height` should be enough.

That said, it might be better to avoid height entirely.
What happens if you add more columns after this conditional formatting?
I think they don't get the conditional formatting applied.
It would be more robust to record the column only.

AndriiSobko replied:
ok, I will avoid it, thanks for the hint.
I think we will add formats after that when all needed records will be created

AndriiSobko replied:
done

rklyne reviewed (changes requested):
Broadly good.
Some bits that need adjusting to make the code more robust.

AndriiSobko committed:
VS-1997

removed conditions

AndriiSobko committed:
VS-1997
We should avoid generating _package here - it's an expensive operation.
self.height should be enough.

AndriiSobko committed:
VS-1997
Added unittests

AndriiSobko committed:
VS-1997
Added unittests

rogerbrandon commented:
Looks good to me. Will await Ronan's view on his comments.
I have had a play locally and exported a file or two, which worked fine.

A 'nice touch' might be to set the column widths.

I am now thinking about how to incorporate what you have added into the Generic Reporting that uses Tablib. The new XLSX format option is easy; it is the other features like headers and footers and formats.

meshy reviewed (changes requested):
While these changes may be valuable additions to the library, I believe that maintaining our own fork of tablib is unwise. The upstream fork is maintained, so we should be opening PRs against it, rather than here. This has several benefits:

1. We will continue getting updates, including security fixes.
1. We will not have to duplicate upstream work to support new versions of Python (etc).
1. We can share these our contributions with the community.

There are a couple of minor things we could do to make this more attractive to upstream. In particular, we should:

1. Open an issue to ask them if they want this new functionality. If they do not feel that it is a worthwhile addition to the project, they may be able to suggest an alternative, complementary, implementation (eg: a plugin).
1. Make the commit messages adhere to [a standard format](https://chris.beams.io/posts/git-commit/). At the moment, they contain our ticket references, some of the messages have been repeated, and the history could do with cleaning up.
